### PR TITLE
Create optimistic report when approving/paying for report with hold expenses

### PR DESCRIPTION
**Pull request by war-in**
### Fixed Issues

$ https://github.com/Expensify/App/issues/41652
PROPOSAL:

### Tests

1. Have report with few expenses (one or two) on hold and the rest (at least one) normal.
2. Enable `Force offline` in `Troubleshooting` (to better see what happens when slow backend replies).
3. Approve only part of the report.
4. See that number of expenses changed for approved report (and held expenses disappeared).
5. See that new report shown at the bottom of the chat with correct expense number and value from one of the expenses.
6. Click on that report and verify that inside are held expenses.
7. Disable `Force offline`.
8. See that new report changed its value to the sum of all held expenses.

### QA Steps

1. Have report with few expenses (one or two) on hold and the rest (at least one) normal.
2. Enable `Force offline` in `Troubleshooting` (to better see what happens when slow backend replies).
3. Approve only part of the report.
4. See that number of expenses changed for approved report (and held expenses disappeared).
5. See that new report shown at the bottom of the chat with correct expense number and value from one of the expenses.
6. Click on that report and verify that inside are held expenses.
7. Disable `Force offline`.
8. See that new report changed its value to the sum of all held expenses.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
  - [x] I added steps for local testing in the `Tests` section
  - [x] I added steps for the expected offline behavior in the `Offline steps` section
  - [x] I added steps for Staging and/or Production testing in the `QA steps` section
  - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
  - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
  - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
  - [x] Android: Native
  - [x] Android: mWeb Chrome
  - [x] iOS: Native
  - [x] iOS: mWeb Safari
  - [x] MacOS: Chrome / Safari
  - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
  - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
  - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
  - [x] I verified that comments were added to code that is not self explanatory
  - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
  - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
  - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
  - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
  - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
  - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
  - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
  - [x] A similar style doesn't already exist
  - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
  - [x] I verified that all the inputs inside a form are aligned with each other.
  - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/44e58a54-b56b-4714-9163-a5d77564064d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/4d8f6a14-413c-4d20-8535-55a762085730

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/9dbd50c7-2d25-40fd-8879-3705efad49df

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/e7ae3f59-ceb5-4a15-bd04-980ba9e117af

</details>

### CODE REVIEW

**Comment by war-in** [comment]
TODO: add `optimisticHoldReportExpenseActionIDs` as param to API

**Review by BartoszGrajdek — APPROVED** [review]

**Review by robertjchen — APPROVED** [review]

**Comment by war-in** [comment]
@ZhenjaHorbach Would be nice to receive a C+ review here :)
Some tests would be welcome too 👍

**Comment by war-in** [comment]
> When we have 4 expenses (2 hold, 2 normal)
And then press approve button the number of elements in a common report decreases, but the amount remains

I'll check on this but the amounts are being updated after the backend response (we optimistically set the value of the new report to the value of the first transaction

**Comment by war-in** [comment]
> Plus, when we turn on the Internet, the new report has a value of 0 for a short period of time (28 seconds in video )

I also reported this in the [slack thread](https://swmansion.slack.com/archives/C049HHMV9SM/p1721820762361279?thread_ts=1714761240.737509&cid=C049HHMV9SM) but @robertjchen found the issue which captures this case https://swmansion.slack.com/archives/C049HHMV9SM/p1721849689066719?thread_ts=1714761240.737509&cid=C049HHMV9SM

**Comment by robertjchen** [comment]
> Plus, when we turn on the Internet, the new report has a value of 0 for a short period of time (28 seconds in video )

Yep! This is known and we'll be addressing this on a followup PR: https://github.com/Expensify/App/issues/43689


> And then press approve button the number of elements in a common report decreases, but the amount remains the same (13 seconds in video )

@war-in Is there another optimistic element that we'll need to account for here? Looks like a preview item? 🤔

**Comment by war-in** [comment]
> @war-in Is there another optimistic element that we'll need to account for here? Looks like a preview item? 🤔

We can decrease the value of the old report by the same value we use for the new one. Let me know if that makes sense

**Comment by robertjchen** [comment]
That works! We can adjust later on if any other corner cases come up 👍

**Review by ZhenjaHorbach — APPROVED** [review]

**Review by robertjchen — APPROVED** [review]

**Comment by war-in** [comment]
Oh, I prepared changes for the old report value 😅

**Comment by war-in** [comment]
I assume we're fine with not changing them then

**Comment by robertjchen** [comment]
@war-in new PR welcome 😄

**Comment by war-in** [comment]
That's only 10 lines of code but I can create it 👍

**Comment by war-in** [comment]
@robertjchen @ZhenjaHorbach could you jump in and verify the [PR](https://github.com/Expensify/App/pull/46178)? 🙏

**Comment by ikevin127** [comment]
Two issues were created coming from this PR:
1. https://github.com/Expensify/App/issues/45902 where the issue was that the offline behaviour did not match the offline behaviour, which was resolved with a simple fix of passing the necessary optimistic data. More details in the selected [proposal](https://github.com/Expensify/App/issues/45902#issuecomment-2252047214).
2. https://github.com/Expensify/App/issues/46371 where the issue was that the necessary data wasn't passed to the `optimisticExpenseReport` and `optimisticExpenseReportPreview` which resulted in the `canUnholdRequest` boolean returning `false`. This was resolved with a simple fix of passing the necessary optimistic data such that the `canUnholdRequest` boolean would return `true` in that specific scenario. More details in the selected [proposal](https://github.com/Expensify/App/issues/46371#issuecomment-2254369728).

**Comment by brunovjk** [comment]
This PR introduced an issue:
#46600  The issue occurred because the app attempted to link to an action that was deleted or moved. This was resolved by clearing the `reportActionID` when the linked action is deleted or moved. More details can be found in the [proposal](https://github.com/Expensify/App/issues/46600#issuecomment-2262047138).

File: src/libs/actions/IOU.ts [hunk 1]
```diff
@@ -6195,6 +6197,203 @@ function getSendMoneyParams(
...
+    failureData: OnyxUpdate[];
+} {
+    const {holdReportActions, holdTransactions} = getHoldReportActionsAndTransactions(iouReport.reportID);
+    const firstHoldTransaction = holdTransactions[0];
+
+    const optimisticExpenseReport = ReportUtils.buildOptimisticExpenseReport(
+        chatReport.reportID,
+        chatReport.policyID ?? iouReport.policyID ?? '',
+        recipient.accountID ?? 1,
+        (firstHoldTransaction?.amount ?? 0) * -1,
```
Comment (s77rt) [inline src/libs/actions/IOU.ts:6241]: This argument is for the total. We should have summed up the hold transactions. (Coming from https://github.com/Expensify/App/issues/48760)

File: src/libs/actions/IOU.ts [hunk 2]
```diff
@@ -6195,6 +6197,203 @@ function getSendMoneyParams(
...
+    });
+
+    const updateHeldTransactions: Record<string, Pick<OnyxTypes.Transaction, 'reportID'>> = {};
+    holdTransactions.forEach((transaction) => {
+        updateHeldTransactions[`${ONYXKEYS.COLLECTION.TRANSACTION}${transaction.transactionID}`] = {
+            reportID: optimisticExpenseReport.reportID,
+        };
+    });
+
+    const optimisticData: OnyxUpdate[] = [
```
Comment (DylanDylann) [inline src/libs/actions/IOU.ts:6295]: Coming from https://github.com/Expensify/App/issues/52696, We need to reset pending action in all places if we use it in optimistic data.

File: src/libs/actions/IOU.ts [hunk 3]
```diff
@@ -6195,6 +6197,203 @@ function getSendMoneyParams(
...
+    optimisticHoldReportID: string;
+    optimisticHoldActionID: string;
+    optimisticHoldReportExpenseActionIDs: OptimisticHoldReportExpenseActionID[];
+    optimisticData: OnyxUpdate[];
+    failureData: OnyxUpdate[];
+} {
+    const {holdReportActions, holdTransactions} = getHoldReportActionsAndTransactions(iouReport.reportID);
+    const firstHoldTransaction = holdTransactions[0];
+
+    const optimisticExpenseReport = ReportUtils.buildOptimisticExpenseReport(
```
[the remark this answers is not in the document]
Comment (situchan) [reply src/libs/actions/IOU.ts:6237]: Here's [another case (when approve/pay the report partially)](https://github.com/Expensify/App/issues/76780) of inconsistent logic between optimistic data and BE.
More details: https://github.com/Expensify/App/issues/76780#issuecomment-3615105177

**Pull request merged**